lambdaclass/mirra_backend
# State deltas for game updates of players and projectiles

> [!warning] 
> Needs to be merged with https://github.com/lambdaclass/champions_of_mirra/pull/2084

## Motivation

Part of #494

## Summary of changes

- Update protobuf
  - Mark fields that we are going to delta as `optional`, this allows to check if there are set or not. Otherwise default value for the type is implied (e.g. 0 for integers)
  - Modify `vertices` to support deltas
  - Add `UNDEFINED` enums to use as `null` value
- Created `diff/2` to receive any 2 terms and output the resulting delta, check the code for a more thorough explanation and edge cases
- Modify creation of update message to handle all cases where something could be `nil` due to the result of diffing
- Send deltas for `obstacles`, `bushes`, and `crates`. This 3 things represented more than 80% of the delta reduction that we could expect to achieve
- `game_client` **is broken due to this change**. Fixing it should be a follow up to this

## How to test it?

Test with this client branch https://github.com/lambdaclass/champions_of_mirra/pull/2084

To test play as usual

We should be doing some load tests and profiling in apps to compare against not doing deltas

## Checklist
